vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -474,6 +474,158 @@ def query_metrics_wrapper(self, query_specs):
         )
         return metrics_values
 
+    def collect_vsan_metrics(self):
+        # type: () -> None
+        self.log.debug("Starting vsan metrics collection (query start time: %s).", self.latest_event_query)
+        latest_metric_time = None
+        collect_start_time = get_current_datetime()
+        try:
+            t0 = Timer()
+            new_health_metrics, new_performance_metrics = self.query_vsan_metrics(
+                collect_start_time - dt.timedelta(hours=2)
+            )
+            self.gauge(
+                'vsphere.vsan.cluster.time',
+                t0.total(),
+                tags=self._config.base_tags,
+                raw=True,
+                hostname=self._hostname,
+            )
+
+            for cluster in new_health_metrics:
+                for metric_name, metric_value in cluster.items():
+                    list_of_tags = ["{}:{}".format(key, value) for key, value in metric_value.items()]
+                    self.gauge(
+                        metric_name,
+                        1,
+                        tags=list_of_tags + self._config.base_tags,
+                        raw=True,
+                        hostname=self._hostname,
+                    )
+
+            for cluster in new_performance_metrics:
+                for entity_type in cluster:
+                    if len(entity_type.value) > 0:
+                        resource_type = entity_type.value[0].metricId.dynamicProperty[0][0]
+                        for given_metric in entity_type.value:
+                            self.log.debug(
+                                "Processing metric with type:%s",
+                                type(given_metric),
+                            )
+                            latest_value = given_metric.values.split(',')[-1]
+                            if latest_value != 'None':
+                                if given_metric.metricId.label in VSAN_PERCENT_METRICS:
+                                    latest_value = float(latest_value) / 100.0
+                                if resource_type == 'cluster':
+                                    if 'vsan.cluster.{}'.format(given_metric.metricId.label) in VSAN_METRICS:
+                                        self.gauge(
+                                            'vsphere.vsan.cluster.{}'.format(given_metric.metricId.label),
+                                            # for now we only collect the latest value
+                                            float(latest_value),
+                                            tags=[
+                                                'vsphere_cluster:{}'.format(given_metric.metricId.dynamicProperty[0][1])
```

Comment:
Could we get the tags from the infrastructure cache? ie `self.infrastructure_cache.get_mor_tags(entity)`

Reply:
I feel like this approach to adding the tags would add unneeded complexity in this situation. Firstly, the infrastructure cache doesn't include disks, so we would still need to use the current approach to get disk tags, creating a "hybrid" approach to adding tags to resources, complicating the code. Secondly, in order to "discover" hosts within clusters and disks within hosts in the vSphereAPI, we already get access to the cluster or host or disk reference already, so just storing their reference's name/id when we already have access to the reference itself is simpler than storing the reference, and then calling for the tags at a later point (when the metric is being gauged).

Reply:
I agree it is a little more complicated/it would be ideal to use the same logic for every resources since the disks are not in the infra cache. But we could still save some work/logic by getting the cluster tags from the cache at line 107. That way, we only get them once, and then we don't need to add them manually to every metric. Additionally, there are more tags in the infra cache than  `vsphere_cluster`, such as `vsphere_type`, `vsphere_folder`, custom tags adding by the user, etc that are useful when querying and filtering the data.